File: Client/client/Assets/Scripts/Engine/Core/Network/MsgDispatcher.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using gtmInterface;
using System;
using gtmEngine.Net;
using FlatBuffers;
using System.Reflection;

namespace gtmEngine
{
    class FlatBufferProcFun<T> : IFlatBufferProcFun where T : struct, FlatBuffers.IFlatbufferObject
    {
        private MsgProcDelegate<T> m_dlg;

        private T m_obj = default(T);

        public FlatBufferProcFun(MsgProcDelegate<T> dlg)
        {
            m_dlg = dlg;
        }

        public override void Invoke(byte[] bytearray)
        {
            FlatBuffers.ByteBuffer buf = new FlatBuffers.ByteBuffer(bytearray);
            m_obj.__init(buf.GetInt(buf.Position) + buf.Position, buf);
            
            try
            {
                m_dlg(m_obj);
            }
            catch (Exception e)
            {
                LogSystem.instance.LogError(LogCategory.GameEngine, "process msg error!" + typeof(T).FullName);
                LogSystem.instance.LogError(LogCategory.GameEngine, e.ToString());
            }
        }
    }

    public class MsgDispatcher : IMsgDispatcher
    {
        #region 变量

        /// <summary>
        /// fb消息句柄
        /// </summary>
        private Dictionary<ulong, IFlatBufferProcFun> m_fbMsgProcDict = new Dictionary<ulong, IFlatBufferProcFun>(100);


        private FlatBuffers.FlatBufferBuilder m_flatBufferBuilder = new FlatBuffers.FlatBufferBuilder(1024);
        /// <summary>
        /// flatBufferBuilder
        /// </summary>
        public override FlatBuffers.FlatBufferBuilder flatBufferBuilder
        {
            get
            {
                m_flatBufferBuilder.Clear();
                return m_flatBufferBuilder;
            }
        }

        #endregion

        #region 继承

        public override void Dispatcher(ulong msgid, byte[] bytearray)
        {
            if (m_MsgType == IMsgType.FlatBuffer)
            {
                DispatcherFbMsg(msgid, bytearray);
            }
        }

        public override void DoClose()
        {
            m_fbMsgProcDict.Clear();
        }

        public override void DoInit()
        {
            
        }

        public override void DoUpdate()
        {
            
        }
        public override void RegisterFBMsg<T>(MsgProcDelegate<T> fbfunc)
        {
            Type type = typeof(T);
            FieldInfo fieldInfo = type.GetField("HashID", BindingFlags.Static | BindingFlags.Public);
            ulong hashid = (ulong)fieldInfo.GetValue(null);

            IFlatBufferProcFun exist;
            if (m_fbMsgProcDict.TryGetValue(hashid, out exist))
            {
                ILogSystem.instance.LogError(LogCategory.GameEngine, "FBMsgProc Exist! " + type.Name);
            }
            else
            {
                m_fbMsgProcDict.Add(hashid, new FlatBufferProcFun<T>(fbfunc));
            }
        }

        public override void UnRegisterFBMsg<T>(MsgProcDelegate<T> fbfunc)
        {
            Type type = typeof(T);
            FieldInfo fieldInfo = type.GetField("HashID");
            ulong hashid = (ulong)fieldInfo.GetValue(null);

            m_fbMsgProcDict.Remove(hashid);
        }

        public override void SendFBMsg(ulong msgid, FlatBufferBuilder builder)
        {
            // 这里做了优化处理，不从flatbuffer里面复制一份数据出来， 而是直接取数据, 减少一次拷贝
            int msgpos = builder.DataBuffer.Position;
            int msglen = builder.DataBuffer.Length - builder.DataBuffer.Position;

            gtmInterface.ByteBuffer buff = new gtmInterface.ByteBuffer();
            UInt16 lengh = (UInt16)(msglen + sizeof(ulong));
            UInt16 biglen = Converter.GetBigEndian(lengh);
            buff.WriteShort(biglen);
            buff.WriteUlong(msgid);
            buff.WriteBytes(builder.DataBuffer.RawBuffer, msgpos, msglen);


            string strout = "";
            for (int i = msgpos; i < builder.DataBuffer.RawBuffer.Length; i++)
            {
                strout = strout + builder.DataBuffer.RawBuffer[i];
            }
            LogSystem.instance.Log(LogCategory.GameEngine, strout);


            if (NetManager.instance != null)
            {
                NetManager.instance.SendMessage(buff);
            }
        }

        private void DispatcherFbMsg(ulong msgid, byte[] bytearray)
        {
            IFlatBufferProcFun procfunc;
            if (m_fbMsgProcDict.TryGetValue(msgid, out procfunc))
            {
                procfunc.Invoke(bytearray);
            }
        }

        #endregion
    }
}













// 这个是从flatbuffer里面复制一份数据出来
//byte[] bytearray = builder.DataBuffer.ToSizedArray();

//gtmInterface.ByteBuffer buff = new gtmInterface.ByteBuffer();
//UInt16 lengh = (UInt16)(bytearray.Length + sizeof(ulong));
//UInt16 biglen = Converter.GetBigEndian(lengh);
//buff.WriteShort(biglen);
//buff.WriteUlong(msgid);
//buff.WriteBytes(bytearray);

File: Client/client/Assets/StreamingAssets/lua/msgdispatcher/msgdispatcher.lua
local flatbuffers = require("3rd/flatbuffers");

---@type netmsg
local netmsg = require 'msgdispatcher/netmsg'

---@class msgdispatcher
msgdispatcher = {

}

msgdispatcher.msgType = {
    flatbuffer = 0,
    protobuf = 1,
}

---@class builder
function msgdispatcher.newBuilder(len)
    return flatbuffers.Builder(len)
end

msgdispatcher.curMsgType = msgdispatcher.msgType.flatbuffer;

-- 消息分发
msgdispatcher.dispatcher = function(msgid, bytearray)

    if msgdispatcher.curMsgType == msgdispatcher.msgType.flatbuffer then
        msgdispatcher.dispatcherFbMsg(msgid, bytearray);
    end
end

-- fb消息注册
msgdispatcher.registerFbMsg = function(msg, handler)
    netmsg.AddListener(msg, handler);
end

-- fb消息取消注册
msgdispatcher.unRegisterFbMsg = function(msg, handler)
    netmsg.RemoveListener(msg, handler);
end

-- 发送fb消息
msgdispatcher.sendFbMsg = function(msg, builder)

    local msgbytes = builder:Output();
    local msgid = msg.HashID;
    local msglen = #msgbytes + 8;

    local bytebuf = global.ByteBuffer();
    msglen = global.Converter.GetBigEndianUInt16(msglen)
    bytebuf:WriteShort(msglen);
    bytebuf:WriteUlong(msgid)
    bytebuf:WriteBytes(msgbytes, 0, #msgbytes)
    global.INetManager:SendMessage(bytebuf);
end

-- fb消息分发
msgdispatcher.dispatcherFbMsg = function(msgid, bytearray)

    local msgbuf = flatbuffers.binaryArray.New(bytearray);

    local eventlib = netmsg.findNetData(msgid);
    if not eventlib then
        return ;
    end

    local msgclass = eventlib.msgclass;
    if not msgclass then
        return ;
    end

    local msg = msgclass.init(msgbuf, 0);
    if not msg then
        return ;
    end

    netmsg.Brocast(msgid, msg);
end

return msgdispatcher;














--local bufAsString = builder:Output();
--local msgid = msg.HashID;
--local strwrite = string.pack("<L", msgid);
--strwrite = strwrite .. bufAsString;
--local package = string.pack("<s2", strwrite)
--local bytebuf = global.ByteBuffer();
--bytebuf:WriteBytes(package, 0, #package);
--global.INetManager:SendMessage(bytebuf);

--@class builder
--msgdispatcher.builder = flatbuffers.Builder(1024)